mmengine/utils/progressbar_rich.py

```diff
@@ -0,0 +1,226 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from collections.abc import Iterable
+from functools import partial
+from multiprocessing import Pool
+from typing import Callable, Tuple
+
+import rich
+from rich.progress import Progress
+
+from ..utils import is_seq_of
+from .timer import Timer
+
+
+class RichProgressBar:
+    """use rich to enhance progress bar.
+
+    This class uses the rich library to enhance the progressbar,
+    Do not instantiate two RichProgressbars at the same time.
+
+    Examples:
+        >>> import mmengine
+        >>> import time
+        >>> bar = mmengine.RichProgressBar()
+        >>> bar.add_task(10)
+        >>> for i in range(10):
+        >>>     bar.update()
+        >>>     time.sleep(1)
+    """
+
+    def __init__(self):
+        self.bar = Progress()
+        self.tasks = []
+        self.descriptions = []
+        self.colors = []
+        self.timer = None
+        self.completed = 0
+        self.infinite = False
+
+    @staticmethod
+    def write(msg: str, color: str = 'blue') -> None:
+        """Write the massage.
+
+        Args:
+            msg (str): Output massage.
+            color (str): text color.
+        """
+        rich.print(f'[{color}]{msg}')
+
+    def add_task(self,
+                 total: int = None,
+                 color: str = 'blue',
+                 description: str = 'Process...') -> int:
+        """Adding tasks to RichProgressbar.
+
+        Args:
+            total (int): Number of total steps, When the input is None,
+                it indicates a task with unknown length. Defaults to None.
+            color (str): Color of progress bar. Defaults to "blue".
+            description (str): Description of progress bar.
+                Defaults to "Process...".
+        Returns:
+            int: added task`s id.
+        """
+        if total is not None:
+            assert not self.infinite, (
+                'The prior task is an infinite task (total is None), '
+                'RichProgressBar can only accept one infinite task')
+
+            if total <= 0:
+                raise ValueError(
+                    'Total only exists if it is greater than zero or None.')
+            self.tasks.append(
+                self.bar.add_task(
+                    f'[{color}]{description}_0/{total}', total=total))
```

Comment:
Why is `_0` here?

Reply:
The 0 here represents the zeroth traversal in the count